### PR TITLE
USHIFT-368: Generalize asset templating

### PR DESCRIPTION
**Pull request by fzdarsky**
Adds a generic function for rendering parameters into the manifest templates of the hosted components and pre-populates the template data with common data such as image pull specs and commonly used MicroShift config parameters.

It then removes the highly repetitive component-specific rendering functions.

Signed-off-by: Frank A. Zdarsky [fzdarsky@redhat.com](mailto:fzdarsky@redhat.com)

Closes [USHIFT-368](https://issues.redhat.com//browse/USHIFT-368)

### CODE REVIEW

File: pkg/assets/applier.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ var (
 
 var assetsEventRecorder events.Recorder = events.NewLoggingEventRecorder("microshift-assets")
 
-type RenderParams map[string]string
+type RenderParams map[string]interface{}
```
Comment (dhellmann) [inline pkg/assets/applier.go:15]: We talked about this offline, but for my recollection later: The parameters to the render function need to be fairly flexible to support different kubeadmin and cert files, so we don't want a full struct type here with a fixed set of fields. We may reexamine that decision in the future, but for now let's stick with the map.

**Review by dhellmann — COMMENTED** [review]
This looks good, but CI is failing with 

```
case.go:256: daemonsets.apps "dns-default" not found
```

**Comment by dhellmann** [comment]
> This looks good, but CI is failing with
> 
> ```
> case.go:256: daemonsets.apps "dns-default" not found
> ```

And when I run the PR, I don't get a DNS container:

```
openshift-ingress          router-default-6795657dbc-qn5pq	 1/1     Running   0          2m17s
openshift-ovn-kubernetes   ovnkube-master-9hlkd                  3/4     Running   0          2m17s
openshift-ovn-kubernetes   ovnkube-node-q5hzs                    1/1     Running   0          2m17s
openshift-service-ca	   service-ca-864554d5b8-jpfxn           1/1     Running   0          2m17s
openshift-storage          topolvm-controller-8479455f95-nxgzf   4/4     Running   0          2m17s
openshift-storage          topolvm-node-mczf4                    4/4     Running   0          111s
```

or namespace

```
$ oc get namespace
NAME                                STATUS   AGE
default                             Active   3m34s
kube-node-lease                     Active   3m36s
kube-public                         Active   3m36s
kube-system                         Active   3m36s
openshift-controller-manager        Active   3m9s
openshift-dns                       Active   3m3s
openshift-infra                     Active   3m9s
openshift-ingress                   Active   3m3s
openshift-kube-controller-manager   Active   3m9s
openshift-ovn-kubernetes            Active   3m3s
openshift-service-ca                Active   3m3s
openshift-storage                   Active   3m3s
```

**Comment by fzdarsky** [comment]
> And when I run the PR, I don't get a DNS container:

Was missing to set one parameter; they looked too similar. Thanks @dhellmann for catching this.

File: pkg/components/components.go [hunk 2]
```diff
@@ -13,7 +13,7 @@ func StartComponents(cfg *config.MicroshiftConfig) error {
 		return err
 	}
 
-	if err := startCSIPLugin(cfg.DataDir + "/resources/kubeadmin/kubeconfig"); err != nil {
+	if err := startCSIPLugin(cfg, cfg.DataDir+"/resources/kubeadmin/kubeconfig"); err != nil {
```
Comment (mangelajo) [inline pkg/components/components.go:16]:
```suggestion
	if err := startCSIPLugin(cfg, filepath.Join(cfg.DataDir, "/resources/kubeadmin/kubeconfig")); err != nil {
```

Comment (mangelajo) [reply pkg/components/components.go:16]: Well this is actually not from your patch; feel free to ignore

**Comment by mangelajo** [comment]
waiting on Doug's final lgtm (not sure if we want to go with the Interface comments as follow up incremental improvement or now)

**Comment by fzdarsky** [comment]
> waiting on Doug's final lgtm (not sure if we want to go with the Interface comments as follow up incremental improvement or now)

There are many other places where path joining needs fixing. I'd suggest doing that in one sweep and IIUC there's already a JIRA issue for that.

**Comment by oglok** [comment]
> I'm fine with merging the PR. It looks good to me and a first step for a clean up. If @dhellmann agrees, we can go ahead and tag it.

@mangelajo just did :-)

**Pull request merged**